src/py/flwr/server/workflow/secure_aggregation/secaggplus_workflow.py

```diff
@@ -0,0 +1,598 @@
+      clients, and hence the number of shares will be determined in the runtime. This
+      allows for dynamic adjustment based on the total number of participating clients.
+    - Similarly, when `reconstruction_threshold` is a float, it is interpreted as the
+      proportion of the number of shares needed for the reconstruction of a private key.
+      This feature enables flexibility in setting the security threshold relative to the
+      number of distributed shares.
+    - `num_shares`, `reconstruction_threshold`, and the quantization parameters
+      (`clipping_range`, `target_quantization_range`, `modulus_range`) play critical
+      roles in balancing privacy, robustness, and efficiency within the SecAgg+
+      protocol.
+    """
+
+    def __init__(  # pylint: disable=R0913
+        self,
+        num_shares: int | float,
+        reconstruction_threshold: int | float,
+        *,
+        max_weight: float = 1000.0,
+        clipping_range: float = 8.0,
+        quantization_range: int = 4194304,
+        modulus_range: int = 4294967296,
+        timeout: float | None = None,
+    ) -> None:
+        self.num_shares = num_shares
+        self.reconstruction_threshold = reconstruction_threshold
+        self.max_weight = max_weight
+        self.clipping_range = clipping_range
+        self.quantization_range = quantization_range
+        self.modulus_range = modulus_range
+        self.timeout = timeout
+
+        self._check_init_params()
+
+    def __call__(self, driver: Driver, context: Context) -> None:
+        """Run the SecAgg+ protocol."""
+        if not isinstance(context, LegacyContext):
+            raise TypeError(
+                f"Expect a LegacyContext, but get {type(context).__name__}."
+            )
+        state = WorkflowState()
+
+        steps = (
+            self.setup_stage,
+            self.share_keys_stage,
+            self.collect_masked_input_stage,
+            self.unmask_stage,
+        )
+        for step in steps:
+            if not step(driver, context, state):
+                return
+
+    def _check_init_params(self) -> None:  # pylint: disable=R0912
+        # Check `num_shares`
+        if not isinstance(self.num_shares, (int, float)):
+            raise TypeError("`num_shares` must be of type int or float.")
+        if isinstance(self.num_shares, int):
+            if self.num_shares <= 2:
+                raise ValueError("`num_shares` as an integer must be greater than 2.")
+            if self.num_shares > self.modulus_range / self.quantization_range:
+                log(
+                    WARN,
+                    "A `num_shares` larger than `mod_range / target_range` "
+                    "will potentially cause overflow when computing the aggregated "
+                    "model parameters.",
+                )
+            if self.num_shares == 1:
+                self.num_shares = 1.0
```

Comment:
will this ever be reached?

Reply:
I think so. This is when users pass `self.num_share=1`, and the int `1` will be converted to float `1.0` here.

Reply:
should we add a test for that?

Reply:
But @panh99 ,given the `self.num_shares <= 2` check above, the one i'm pointing to here is not going to be evaluated ever because a `ValueError` is going to be triggered.

Reply:
Oh I get it now! Thanks, I will fix it.